src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -0,0 +1,194 @@
+/**
+ * @file kde_rules_impl.hpp
+ * @author Roberto Hueso (robertohueso96@gmail.com)
+ *
+ * Implementation of rules for Kernel Density Estimation with generic trees.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_KDE_RULES_IMPL_HPP
+#define MLPACK_METHODS_KDE_RULES_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "kde_rules.hpp"
+
+namespace mlpack {
+namespace kde {
+
+template<typename MetricType, typename KernelType, typename TreeType>
+KDERules<MetricType, KernelType, TreeType>::KDERules(
+    const arma::mat& referenceSet,
+    const arma::mat& querySet,
+    arma::vec& densities,
+    const double relError,
+    const double absError,
+    const std::vector<size_t>& oldFromNewQueries,
+    MetricType& metric,
+    KernelType& kernel) :
+    referenceSet(referenceSet),
+    querySet(querySet),
+    densities(densities),
+    absError(absError),
+    relError(relError),
+    oldFromNewQueries(oldFromNewQueries),
+    metric(metric),
+    kernel(kernel),
+    lastQueryIndex(querySet.n_cols),
+    lastReferenceIndex(referenceSet.n_cols),
+    baseCases(0),
+    scores(0)
+{
+  // Nothing to do.
+}
+
+//! The base case.
+template<typename MetricType, typename KernelType, typename TreeType>
+inline force_inline
+double KDERules<MetricType, KernelType, TreeType>::BaseCase(
+    const size_t queryIndex,
+    const size_t referenceIndex)
+{
+  double distance = metric.Evaluate(querySet.col(queryIndex),
+                                    referenceSet.col(referenceIndex));
+  if (tree::TreeTraits<TreeType>::RearrangesDataset)
+    densities(oldFromNewQueries.at(queryIndex)) += kernel.Evaluate(distance);
```

Comment:
I'm undecided on this one.  I think it may be quicker to do `densities(queryIndex) = ...` here, and then re-map all the indices after the computation is done.  However that is just my intuition, and that may not actually provide any speedup for the KDE problem.  Do you have any thoughts here?

Reply:
I think the issue would be more that we would be accessing `densities()` in a mostly linear order which can really help with caching.  So `BaseCase()` is likely to be called multiple times with adjacent points... i.e. if we are at some node combination we'll loop over all query points and all reference points and call `BaseCase()`, and often these points will be sequential.

Like I said my intuition could be wrong here---it might not make much of a speed difference---but it's probably worth checking. :)

Reply:
After testing this there's a slight speedup so it's implemented in bd2e970568452930cb2390a83581e4d80710383d :) Thanks for the idea!